curvefs/src/metaserver/recycle_cleaner.cpp

```diff
@@ -0,0 +1,382 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2022-08-25 15:39:21
+ * @Author: chenwei
+ */
+
+#include "curvefs/src/metaserver/recycle_cleaner.h"
+
+#include <list>
+#include <vector>
+
+using curve::client::MetaServerOption;
+using curve::common::StringToUll;
+using curvefs::client::common::AddUllStringToFirst;
+using curvefs::client::rpcclient::MDSBaseClient;
+using curvefs::client::rpcclient::MdsClientImpl;
+using curvefs::mds::FsInfo;
+using curvefs::mds::FSStatusCode;
+using curvefs::mds::topology::PartitionTxId;
+
+namespace curvefs {
+namespace metaserver {
+bool RecycleCleaner::UpdateFsInfo() {
+    uint32_t fsId = partition_->GetFsId();
+    FsInfo fsInfo;
+    FSStatusCode ret = mdsClient_->GetFsInfo(fsId, &fsInfo);
+    if (ret != FSStatusCode::OK) {
+        LOG(WARNING) << "GetFsInfo failed, ret = " << FSStatusCode_Name(ret)
+                     << ", fsId = " << fsId;
+        return false;
+    }
+
+    fsInfo_ = fsInfo;
+    VLOG(1) << "UpdateFsInfo, fsId = " << fsId
+            << ", recycle time = " << GetRecycleTime();
+    return true;
+}
+
+bool RecycleCleaner::IsDirTimeOut(const std::string& dir) {
+    uint32_t hour = GetRecycleTime();
+    if (hour == 0) {
+        return true;
+    }
+
+    time_t timeNow;
+    time(&timeNow);
+
+    struct tm tmDir;
+    memset(&tmDir, 0, sizeof(tmDir));
+    char* c = strptime(dir.c_str(), "%Y-%m-%d-%H", &tmDir);
+
+    time_t dirTime = mktime(&tmDir);
+    if (dirTime <= 0) {
+        LOG(WARNING) << "convert dir to time fail, dir = " << dir;
+        return false;
+    }
+
+    if (timeNow > dirTime + hour * 3600) {
+        return true;
+    }
+
+    return false;
+}
+
+bool RecycleCleaner::DeleteNode(const Dentry& dentry) {
+    uint32_t fsId = partition_->GetFsId();
+    uint64_t parent = dentry.parentinodeid();
+    uint64_t inodeid = dentry.inodeid();
+    std::string name = dentry.name();
+    FsFileType type = dentry.type();
+    LOG(INFO) << "RecycleCleaner DeleteNode, " << dentry.ShortDebugString();
+    // Code refers to the implementation of fuse_client.cpp DeleteNode()
+    // 1. delete dentry
+    auto ret = metaClient_->DeleteDentry(fsId, parent, name, type);
```

Comment:
Whether `ret` here can determine whether the current state is the leader

Reply:
The process of deletenode here is consistent with that of fuseclient. Whether to consider using the fuseclient function directly.

Reply:
> Whether ret here can determine whether the current state is the leader
The request send to the leader by metaclient.
There is no need to pay attention to whether it is a leader or not.

> The process of deletenode here is consistent with that of fuseclient. Whether to consider using the fuseclient function directly.

There are many things in the implementation of the client that are not needed, such as some caches